internal/migrate/staging_service.go

```diff
@@ -0,0 +1,283 @@
+	for contractLocation, res := range s.stageContracts(ctx, validContracts) {
+		res.WasValidated = true
+		results[contractLocation] = res
+	}
+
+	return results, nil
+}
+
+func (s *stagingServiceImpl) stageContracts(ctx context.Context, contracts []stagedContractUpdate) map[common.AddressLocation]stagingResult {
+	results := make(map[common.AddressLocation]stagingResult)
+	for _, contract := range contracts {
+		txId, err := s.stageContract(
+			ctx,
+			contract,
+		)
+		if err != nil {
+			results[contract.DeployLocation] = stagingResult{
+				Err:  fmt.Errorf("failed to stage contract: %w", err),
+				TxId: txId,
+			}
+		} else {
+			results[contract.DeployLocation] = stagingResult{
+				Err:  nil,
+				TxId: txId,
+			}
+		}
+	}
+
+	return results
+}
+
+func (s *stagingServiceImpl) stageContract(ctx context.Context, contract stagedContractUpdate) (flow.Identifier, error) {
+	s.logger.StartProgress(fmt.Sprintf("Staging contract %s", contract.DeployLocation))
+	defer s.logger.StopProgress()
+
+	// Check if the staged contract has changed
+	if !s.hasStagedContractChanged(contract) {
+		return flow.EmptyID, nil
+	}
+
+	cName := cadence.String(contract.DeployLocation.Name)
+	cCode := cadence.String(contract.Code)
+
+	// Get the account for the contract
+	account, err := s.state.Accounts().ByAddress(flow.Address(contract.DeployLocation.Address))
+	if err != nil {
+		return flow.Identifier{}, fmt.Errorf("failed to get account for contract %s: %w", contract.DeployLocation.Name, err)
+	}
+
+	tx, _, err := s.flow.SendTransaction(
+		context.Background(),
+		transactions.SingleAccountRole(*account),
+		flowkit.Script{
+			Code: templates.GenerateStageContractScript(MigrationContractStagingAddress(s.flow.Network().Name)),
+			Args: []cadence.Value{cName, cCode},
+		},
+		flow.DefaultTransactionGasLimit,
+	)
+	if err != nil {
+		return flow.Identifier{}, err
+	}
+
+	return tx.ID(), nil
+}
+
+func (s *stagingServiceImpl) hasStagedContractChanged(contract stagedContractUpdate) bool {
+	// Get the staged contract code
+	stagedCode, err := getStagedContractCode(context.Background(), s.flow, contract.DeployLocation)
+	if err != nil {
+		// swallow error, if we can't get the staged contract code, we should stage
+		return true
+	}
+
+	if stagedCode == nil {
+		return true
+	}
+
+	// If the staged contract code is different from the contract code, we need to stage it
+	if string(stagedCode) != string(contract.Code) {
+		return true
+	}
+
+	return false
+}
+
+func (s *stagingServiceImpl) convertToStagedContracts(contracts []*project.Contract) ([]stagedContractUpdate, error) {
+	// Collect all staged contracts
+	stagedContracts := make([]stagedContractUpdate, 0)
+	for _, contract := range contracts {
+		rawScript := flowkit.Script{
+			Code:     contract.Code(),
+			Location: contract.Location(),
+			Args:     contract.Args,
+		}
+
+		// Replace imports in the contract
+		script, err := s.flow.ReplaceImportsInScript(context.Background(), rawScript)
```

Comment:
Did you try the program method for replacing imports?

Reply:
Yeah, I looked at that one.  It looks like it's for the other way around (e.g. `AddressLocation` -> `StringLocation`, not vice versa)